src/libmongoc/src/mongoc/mongoc-cluster.c

```diff
@@ -612,6 +600,18 @@ mongoc_cluster_run_command_monitored (mongoc_cluster_t *cluster,
       mongoc_apm_command_failed_cleanup (&failed_event);
    }
 
+   if (retval && _mongoc_cse_is_enabled (cluster->client)) {
```

Comment:
If I understand correctly, the two changes here are:

 * Moving `_mongoc_cse_auto_decrypt` until _after_ all APM callbacks have been dispatched
 * Only decrypting if the original command was successful

> Logic try to decrypt an error reply was considered. I do not think an error reply can contain ciphertext to decrypt.

Just throwing ideas here, but a possible edge case might be if an error message included some document data (e.g. duplicate value for unique index, JSON schema violation); however, I don't think that's important to handle.

Reply:
>If I understand correctly, the two changes here are:
>
> * Moving _mongoc_cse_auto_decrypt until after all APM callbacks have been dispatched
> * Only decrypting if the original command was successful

Correct.

> Just throwing ideas here, but a possible edge case might be if an error message included some document data (e.g. duplicate value for unique index, JSON schema violation); however, I don't think that's important to handle.

Good idea. A duplicate value for a unique index could include ciphertext in the response. Here is an example with the shell:

```js
db.encryptedColl.createIndex({encrypted: 1}, {unique: true})
db.runCommand({insert: "encryptedColl", documents: [{encrypted: "ciphertext"}]})
got = db.runCommand({insert: "encryptedColl", documents: [{encrypted: "ciphertext"}]})
printjson (got)
/* Prints:
{
    "n" : 0,
    "electionId" : ObjectId("7fffffff0000000000000001"),
    "opTime" : {
            "ts" : Timestamp(1654288999, 3),
            "t" : NumberLong(1)
    },
    "writeErrors" : [
            {
                    "index" : 0,
                    "code" : 11000,
                    "errmsg" : "E11000 duplicate key error collection: test.encryptedColl index: encrypted_1 dup key: { encrypted: \"ciphertext\" }",
                    "keyPattern" : {
                            "encrypted" : 1
                    },
                    "keyValue" : {
                            "encrypted" : "ciphertext"
                    }
            }
    ],
    "ok" : 1,
    "$clusterTime" : {
            "clusterTime" : Timestamp(1654289001, 1),
            "signature" : {
                    "hash" : BinData(0,"AAAAAAAAAAAAAAAAAAAAAAAAAAA="),
                    "keyId" : NumberLong(0)
            }
    },
    "operationTime" : Timestamp(1654288999, 3)
}
*/
```

However, trying to decrypt an error reply may also error. Rather than overwrite the original command error, I think it is preferable to return the command error immediately.